execute-async-function.js

```diff
@@ -76,6 +76,62 @@
         });
     }
 
+    /** Create a promise that resolves when chrome.tabs.onUpdated fires with the id
+     * @param {string} id ID for the tab we're expecting.
+     * Tabs without the ID will not resolve or reject this promise.
+     * @returns {Promise} Promise that resolves when chrome.tabs.onUpdated.addListener fires. */
+    function promisifyTabUpdate(id) {
+
+        let mainPromise = new Promise((resolve, reject) => {
+            const tabUpdatedListener = (tabId, changeInfo, tab) => {
+                // The onUpdated event is called multiple times during a single load.
+                // the status of 'complete' is called only once, when it is finished.
+                if (tabId === id && changeInfo.status === 'complete') {
+                    removeListeners();
+                    resolve({tabId:tabId, changeInfo:changeInfo, tab:tab});
+                }
+            };
+
+            // This will happen when the tab or window is closed before it finishes loading
+            const tabRemovedListener = (tabId, removeInfo) => {
+                if (tabId === id) {
+                    removeListeners();
+                    reject(new Error(`The tab with id = ${tabId} was removed before it finished loading.`));
+                }
+            }
+
+            // This will happen when the tab is replaced.  This is untested, not sure how to recreate it.
+            const tabReplacedListener = (addedTabId, removedTabId) => {
+                if (removedTabId === id) {
+                    removeListeners();
+                    reject(new Error(`The tab with id = ${removedTabId} was replaced before it finished loading.`));
+                }
+            }
+
+            const removeListeners = () => {
+                chrome.tabs.onUpdated.removeListener(tabUpdatedListener);
+                chrome.tabs.onRemoved.removeListener(tabRemovedListener);
+                chrome.tabs.onReplaced.removeListener(tabReplacedListener);
+            }
+
+            chrome.tabs.onUpdated.addListener(tabUpdatedListener);
+            chrome.tabs.onRemoved.addListener(tabRemovedListener);
+            chrome.tabs.onReplaced.addListener(tabReplacedListener);
+        });
+
+        // Although I have onRemoved and onReplaced events watching to reject the promise,
+        // there is nothing in the chrome extension api documentation that guarantees this will be an exhaustive approach.
+        // So to account for the unknown, I am adding an auto-timeout feature to reject the promise after 2 minutes.
+        let timeoutPromise = new Promise ( (resolve, reject) => {
+            const secondsToTimeout = 120;
+            setTimeout(() => {
+                reject(new Error(`The tab loading timed out after ${secondsToTimeout} seconds.`));
+            }, secondsToTimeout * 1000);
+        });

```

Comment:
@mogelbuster This idea makes good sense, but it should be opt-in (or at least opt-out).

Could we add an optional timeout (in ms, as everything else in Google's API is in ms) to `createAndWait`/`reloadAndWait`, maybe with a default of `12e4`, then only add the reject timeout if `timeout > 0`?

Reply:
Great suggestion!  Check out my latest commit, I included that functionality and updated the documentation for the readme and the function definitions appropriately.  Let me know if anything needs to be changed.